moveit_ros/planning/planning_request_adapter_plugins/src/fix_start_state_collision.cpp

```diff
@@ -115,37 +127,106 @@ class FixStartStateCollision : public planning_request_adapter::PlanningRequestA
-      random_numbers::RandomNumberGenerator& rng = prefix_state->getRandomNumberGenerator();
+      moveit::core::RobotStatePtr original_state(new moveit::core::RobotState(start_state));
+      random_numbers::RandomNumberGenerator& rng = original_state->getRandomNumberGenerator();
 
       const std::vector<const moveit::core::JointModel*>& jmodels =
           planning_scene->getRobotModel()->hasJointModelGroup(req.group_name) ?
               planning_scene->getRobotModel()->getJointModelGroup(req.group_name)->getJointModels() :
               planning_scene->getRobotModel()->getJointModels();
 
+      // Record the cartesian pose of every link to ensure no sudden movements will occur.
+      // The "furthest" links are checked first, as their movement is the largest.
+
+      // Check if an end effector link is set that is not part of the robot
+      const moveit::core::LinkModel* ee_link_model;
+      if (!req.goal_constraints.empty())
+      {
+        if (!req.goal_constraints[0].position_constraints.empty())
+          ee_link_model = start_state.getLinkModel(req.goal_constraints[0].position_constraints[0].link_name);
+        else if (!req.goal_constraints[0].orientation_constraints.empty())
+          ee_link_model = start_state.getLinkModel(req.goal_constraints[0].orientation_constraints[0].link_name);
+      }
+      else
+      {
+        ROS_ERROR("No goal constraint set; aborting FixStartStateCollision.");
+        return planner(planning_scene, req, res);
+      }
+
+      // Collect the links that will be checked
+      std::vector<const moveit::core::LinkModel*> link_models;
+      link_models.resize(jmodels.size());
+      for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+        link_models[i] = jmodels[i]->getChildLinkModel();
+
+      if (std::find(link_models.begin(), link_models.end(), ee_link_model) == link_models.end())
+      {  // If the end effector is a separate link, add it to the front of the vector
+        link_models.resize(jmodels.size() + 1);
+        link_models[0] = ee_link_model;
+        for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+          link_models[i + 1] = jmodels[i]->getChildLinkModel();
+      }
+
+      // Store the unchanged link positions
+      EigenSTL::vector_Isometry3d original_link_poses;
+      original_link_poses.resize(link_models.size());
+      Eigen::Isometry3d new_link_pose;
+
+      for (std::size_t i = 0; i < link_models.size(); ++i)
+        original_link_poses[i] = start_state.getGlobalLinkTransform(link_models[i]);
+
+      double dist, max_dist;
+      double local_jiggle_fraction = jiggle_fraction_;
       bool found = false;
-      for (int c = 0; !found && c < sampling_attempts_; ++c)
+      for (int c = 0; !found && c < max_sampling_attempts_; ++c)
       {
+        // Change each joint slightly ("jiggle")
         for (std::size_t i = 0; !found && i < jmodels.size(); ++i)
```

Comment:
`found` doesn't need to be checked in this loop as it isn't changed by the loop.